### PR TITLE
Enable NI_Vector256_Create on AVX1

### LINKED ISSUE #72506: Vector256.Create() randomly throws PlatformNotSupportedException
### Description

Calling `Vector256.Create()` on a machine with no support for AVX2 will throw `PlatformNotSupportedException` after a short amount of time.

### Reproduction Steps

```cs
using System.Runtime.Intrinsics;

int i = 0;
while (true) {
    Vector256.Create(0f);
    if (i++ % 10000 == 0) Console.WriteLine("OK " + i);
}
```

### Expected behavior

The program above should run indefinitely.

### Actual behavior

The program crashes after a short amount of time:

```
...
OK 3390001
OK 3400001
Unhandled exception. System.PlatformNotSupportedException: Operation is not supported on this platform.
   at System.Runtime.Intrinsics.Vector256.Create(Single value)
   at Program.<Main>$(String[] args) in ...\Program.cs:line 4
```

### Regression?

I'm pretty sure it is. Changing the target to .NET 6 works as expected, and I've only found this issue after upgrading to .NET 7 preview 6.

### Known Workarounds

_No response_

### Configuration

.NET SDK: 7.0.100-preview.6.22352.1 (commit 492644e08e)
OS: Windows 10 (6.3.9600, x64)
CPU: Intel i3 Gen2 (Sandy Bridge, with no AVX2 support)

### Other information

_No response_

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes https://github.com/dotnet/runtime/issues/72506

### CODE REVIEW

File: src/coreclr/jit/hwintrinsic.cpp [hunk 1]
```diff
@@ -319,6 +319,10 @@ NamedIntrinsic HWIntrinsicInfo::lookupId(Compiler*         comp,
     // ISA is unsupported. For Vector256 this is when AVX2 is unsupported since integer types
     // can't get properly accelerated.
 
+
+    // We support some Vector256 intrinsics on AVX-only CPUs
+    bool isLimittedVector256Isa = false;
```
Comment (teo-tsirpanis) [inline src/coreclr/jit/hwintrinsic.cpp:324]: Typo, "limitted" -> "limited".

**Comment by EgorBo** [comment]
@tannergooding PTAL

**Comment by tannergooding** [comment]
An alternative and likely better fix would to just remove or update the recursive path from the `Create` calls. Such a path is only used for indirect invocation (reflection, delegates, function pointers, etc) and its not really worth optimizing them when the software fallback already does the right thing. Additionally, it simplifies the handling and reduces risk of other errors due to future ISA or implementation changes.

**Comment by tannergooding** [comment]
> @tannergooding I am not sure I understand how it's better - if we remove the recursive pass we won't be able to optimize Vector256.Create on AVX1 so e.g. Avx.Xor(Vector256.Create(...), Vector256.Create(...)) will be terribly slow.

AVX only hardware is pretty rare (only `Sandy Bridge` and `Ivy Bridge` processors) and performance is already going to be problematic in various cases due to other APIs and scenarios not being accelerated.

We also continue hitting various edge cases and hard to find bugs due to the complex relationship between what's available on AVX vs what requires AVX2. If we want to provide `AVX` only acceleration then that should likely be looked at more broadly and we should consider the best way to model that to ensure that all the "expected" functions work as users require. It's notably too late in the .NET 7 cycle to do that, however, and is decently non-trivial since it requires looking at each `Vector256.*` ISA and considering when it can light up for each of the supporting types (often `float`/`double` being AVX and `integer` types being `AVX2`; but there is some intermixing there).

Keeping this so that `Vector256.Create` is only accelerated when `Vector256.IsHardwareAccelerated` returns `true` simplifies the overall handling and ensures that .NET 7 ships more robustly. This will not impact cloud providers (who, to my knowledge, are all on Broadwell or later) and won't impact most end users.

**Comment by tannergooding** [comment]
Taking this change requires running the outerloop ISA tests and validating that no AVX2 instructions are generated by any path that `Vector256.Create` can hit, including under `GT_CNS_VEC` and any other tree shapes that morph, valuenum, lowering, or other phases may introduce.

**Comment by EgorBo** [comment]
@tannergooding I think it's better to do both drop the recursive calls (I've just done it) and still accelerate Vector256.Create for AVX1 (also done).

**Comment by tannergooding** [comment]
> and still accelerate Vector256.Create for AVX1 (also done).

This is still going to require running the outerloop ISA stress jobs and confirming that other phases aren't expecting to only be hit when `AVX2` is enabled.

**Comment by EgorBo** [comment]
Because it's possible to work with AVX-only in .NET 6.0 via Vector256.Create + Avx.X APIs - I don't see why we want to break it

**Comment by EgorBo** [comment]
> > and still accelerate Vector256.Create for AVX1 (also done).
> 
> This is still going to require running the outerloop ISA stress jobs and confirming that other phases aren't expecting to only be hit when `AVX2` is enabled.

do you mean `runtime outerloop` pipeline?

**Comment by tannergooding** [comment]
> Because it's possible to work with AVX-only in .NET 6.0 via Vector256.Create + Avx.X APIs - I don't see why we want to break it

It's semi-possible and you run into some poor experiences and bad codegen because the JIT can't optimize everything as it needs. We also hit and had to patch several bugs related to mismatches.

Generating AVX only code (when AVX2 isn't supported) has been one of the biggest issues we've faced because the `Vector<T>` never had such support and various paths don't expect it. The raw hardware intrinsics mostly mitigate this in that they are directly tied to an ISA/base type. The helper APIs (like `Create` or `GetElement`) are not directly tied to an ISA or base type, they are generic and you have to keep in mind in the JIT exactly how each tree shape is handled in the backend.

**Comment by tannergooding** [comment]
/azp run runtime-coreclr jitstress-isas-x86

^ this one

**Comment by EgorBo** [comment]
@tannergooding sure, I understand your point, but as we now have stronger AOT-only story (NativeAOT) we should be ready for cases when developers limit capabilities to some lower common denominator e.g. `--instruction-set avx` (even some modern games distribute binaries for avx1 only afaik). So while for JIT ivy-bridge like machines might be rare - it's not the case for AOT case like ^.

**Comment by tannergooding** [comment]
Sure, the main concern here is destabilizing .NET 7 and the work that might be required to ensure that everything end to end works as expected particularly with all the changes and issues that had cropped up.

**Comment by saucecontrol** [comment]
> Keeping this so that Vector256.Create is only accelerated when Vector256.IsHardwareAccelerated returns true simplifies the overall handling and ensures that .NET 7 ships more robustly. This will not impact cloud providers (who, to my knowledge, are all on Broadwell or later) and won't impact most end users.

This may not impact Azure or AWS, but some of the smaller hosts definitely have these older machines in the mix.  I have a client running .NET apps on a large number of Rackspace Cloud VMs, and all the ones I saw in a recent spot check were Sandy Bridge ([Xeon E5-2680](https://ark.intel.com/content/www/us/en/ark/products/64583/intel-xeon-processor-e52680-20m-cache-2-70-ghz-8-00-gts-intel-qpi.html) to be precise).

`Vector256.Create(float)` has been accelerated since 3.0 on AVX-only, and no matter how uncommon that hardware is, disabling acceleration now would be a nasty surprise for anyone that's running currently-fast AVX code paths that would be crippled on update to 7.0.

**Comment by EgorBo** [comment]
@tannergooding ping

**Comment by tannergooding** [comment]
> Vector256.Create(float) has been accelerated since 3.0 on AVX-only, and no matter how uncommon that hardware is, disabling acceleration now would be a nasty surprise for anyone that's running currently-fast AVX code paths that would be crippled on update to 7.0.

More generally speaking, I'm raising a concern around the number of bugs we've had and continue hitting on trying to split AVX/AVX2 functionality.

There is complexity in ensuring that support exists, that it works well, and that some code path doesn't accidentally pessimize it because it really needs some functionality that was only in AVX2 instead (not to mention that there isn't a bug, of which we've had a few, where AVX2 is generated when only AVX is supported). This is why when AMD, Intel, LLVM, and others got together to define the `x86-x64-v*` baselines, they were:
* v1: CMOV, CMPXCHG8B, SSE, SSE2
* v2: ..., CMPXCHG16B, POPCNT, SSE3, SSSE3, SSE4.1, SSE4.2
* v3: ..., AVX, AVX2, BMI1, BMI2, FMA, F16C, LZCNT, MOVBE
* v4: ..., AVX512F, AVX512BW, AVX512CD, AVX512DQ, AVX512VL

In order to keep things "simple" and avoid explosions around support matrices. Supporting just "SSSE3" or just "AVX" doesn't necessarily make sense and so `Vector256.IsHardwareAccelerated` reports "false" when only AVX is supported. And while I agree that not regressing an existing scenario of `Vector256.Create(float)` working on AVX only is goodness, there is a balance in ensuring that the overall support in the JIT remains robust/correct and that we don't unnecessarily complicate the JIT implementation.

Some examples of where this gets complex, even for `Create(float)` are that `VBROADCASTSS ymm, xmm` is AVX2; AVX only exposes `VBROADCASTSS ymm, mem32`. The same goes for `VPERMPS`. You then get into more complex cases with integer types where `Vector256<int>` ends up partially accelerated because `GetLower`/`GetUpper` and `WithLower`/`WithUpper` can be "always accelerated" as can `Load`, `Store`, and `Test`. But other operations can't be. So it becomes a question of "where do we draw the line" and how complex do we make the import checks around whether something can or cannot be intrinsified in importation.

**Comment by EgorBo** [comment]
> So it becomes a question of "where do we draw the line" and how complex do we make the import checks around whether something can or cannot be intrinsified in importation.
 
@tannergooding I agree but It still reports `Vector256.IsHardwareAccelerated` as false but significantly improves perf for the most common operations Vector256.Create and fixes the unexpected recursion = PlatformNotSupported. Other less common portable APIs can be accelerated as well as we hit them, but they not that common as Create IMO

**Comment by tannergooding** [comment]
> I do think we need to fix https://github.com/dotnet/runtime/issues/72506 for 7.0

Yes, 100% agree

> there are no good reasons to break what worked in 3.1-6.0

Provided we have sufficient coverage and validation that it is fixed and isn't regressing other key scenarios, and isn't going to be something we're likely to decide to regress in the future due to the complexity required in the JIT, etc.

**Comment by EgorBo** [comment]
@tannergooding so can you approve it?
the jit change is literally few lines of code, CI is happy
the rest is just a clean up to remove the hack that is only improving perf for indirect usages (which are slow by definition)

**Review by tannergooding — APPROVED** [review]

**Pull request merged**